In [59]:
import pandas as pd
df = pd.read_csv('LLCP2024.ASC', sep='\s+', header=None, on_bad_lines='skip')
print(df.info())
print(df.describe())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90522 entries, 0 to 90521
Data columns (total 45 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       90522 non-null  int64 
 1   1       90522 non-null  int64 
 2   2       90522 non-null  int64 
 3   3       90522 non-null  int64 
 4   4       90522 non-null  int64 
 5   5       90522 non-null  int64 
 6   6       90522 non-null  int64 
 7   7       90522 non-null  int64 
 8   8       90522 non-null  int64 
 9   9       90522 non-null  int64 
 10  10      90522 non-null  object
 11  11      90522 non-null  object
 12  12      90522 non-null  object
 13  13      90522 non-null  object
 14  14      90522 non-null  object
 15  15      90522 non-null  object
 16  16      90522 non-null  object
 17  17      90522 non-null  object
 18  18      90522 non-null  object
 19  19      90522 non-null  object
 20  20      90522 non-null  object
 21  21      90522 non-null  object
 22  22      90522 non-null

In [60]:
df.shape

(90522, 45)

In [61]:
df.columns


Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44],
           dtype='int64')

In [62]:
df


,0,1,2,3,4,5,6,7,8,9,...,35,36,37,38,39,40,41,42,43,44
0,1,202282024,11002024000001,11,12101,2,2,3028888032211,11,2222,...,12119121121131,01,211112122786064163059422249211291,1,1,2,411,02,62,9200010000011220.0
1,1,202212024,11002024000002,11,1210211,1,18888,31211,11,2122,...,10.46529427440.298009,307.169688,11119121111131,01,21111113280607017808165258332147,1,2,3110270200202010000020535249200010000011121,NaN,NaN
2,1,202212024,11002024000003,11,1210311,1,2308801013141,42,2222,2222213,...,1311112,221131,01,21111108159507819808845225321139,2323333333121,015024500750034,114211002014001,21,NaN,NaN
3,1,202212024,11002024000005,11,1210311,1,38888,51212,11,2222,...,1111122,121132,01,21111106147406817305897197721132,2323333333411,01,623920001000001,21,NaN,NaN
4,1,202212024,11002024000006,11,1210211,1,3078888012211,18,2222,212222144,...,1211111,121132,01,21111107154406717009662333642126,1123333331411,02,623920001000001,21,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
90517,78,707272024,12002024001333,11111,1,1,2011,38888,881211,71,...,1.0,1.0,1041.0,38888.0,12211.0,38.0,2222,2222223,543,1210377
90518,78,707142024,12002024001336,11111,1,1,2011,4880388883212,41,2222,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
90519,78,707232024,12002024001337,11111,1,1,2041,4113088103211,11,2212,...,1.0,1.0,1052.0,5012088772221.0,22.0,2222.0,2222223,541,1278877,013605042222222
90520,78,707202024,12002024001339,11111,1,1,2031,4012504011211,17,222112222223,...,1.0,1.0,2021.0,38888.0,101121.0,18.0,2222,1222223,161,1210107


### PROBLEM STATEMENT


Cardiovascular disease—including heart disease and heart attack—is a growing public health crisis in Kenya, driven by rapid urbanization, dietary shifts toward processed foods, rising obesity, increasing tobacco use, and limited access to preventive screening in many communities. Kenya's Ministry of Health and county health departments need reliable, interpretable signals to identify high-risk populations and allocate scarce prevention resources; healthcare providers in Kenyan hospitals, clinics, and community health units need simple, evidence-based tools to support opportunistic screening during routine visits; health systems such as NHIF/SHA and private insurers need data-driven approaches to design preventive wellness programs and reduce the growing burden of non-communicable disease (NCD) admissions; and Kenyan patients and community health volunteers (CHVs) need understandable insights that translate into actionable lifestyle changes within the local context.

While Kenya does not yet have a population-scale behavioral risk factor survey comparable in size and variable richness to the U.S. CDC BRFSS, the risk factors captured by BRFSS—smoking, obesity/BMI, physical inactivity, diabetes, hypertension, high cholesterol, poor general health, and inadequate sleep—are the same modifiable risk factors driving cardiovascular disease in Kenya (as documented by Kenya's National NCD Strategic Plan and the WHO STEPwise surveys). This project therefore uses the CDC BRFSS 2024 raw public dataset as a methodological proxy and proof-of-concept: the goal is to develop and validate an interpretable machine learning pipeline and deployable screening tool that could be adapted for Kenyan populations once equivalent local survey data (e.g., from the Kenya STEPwise survey, Kenya Demographic and Health Survey, or a future Kenya BRFSS-style instrument) becomes available at sufficient scale.

Specifically, this project will:

Build and evaluate an interpretable machine learning model using CDC BRFSS 2024 data that estimates an individual's risk of having heart disease or a prior heart attack (combined outcome) based on behavioral and clinical self-report indicators.
Identify and rank the strongest contributing risk factors—focusing on those most relevant to the Kenyan NCD burden (smoking, BMI/obesity, physical inactivity, diabetes, hypertension).
Deploy a screening-style web application where a user can input risk factors and receive a risk probability, a high/low risk classification, and an explanation of the top contributing factors—designed so that the interface, variable set, and explanation approach could be directly reused with Kenyan survey data when available.
Discuss transferability: explicitly analyze which BRFSS-derived findings are likely to generalize to Kenyan populations and which may not, citing differences in demographics, healthcare access, dietary patterns, and genetic/environmental context.
Disclaimer: This tool is built on U.S. survey data and is intended for research, education, and proof-of-concept purposes only. It is not validated for clinical use in Kenya and should not be used for individual medical decisions. Adaptation to the Kenyan context would require local data collection, recalibration, and clinical validation.

Stakeholders (Kenya-contextualized)
Stakeholder	Need
Kenya Ministry of Health / NCD Division	Evidence-based tools to prioritize cardiovascular prevention in national NCD strategy
County health departments (47 counties)	Identify high-risk communities for targeted screening and outreach
Healthcare providers (public hospitals, clinics, community health units)	Simple screening support during routine visits, especially in resource-limited settings
Community Health Volunteers (CHVs)	Easy-to-use risk communication tools for household visits
NHIF / SHA and private health insurers	Data-driven design of preventive wellness benefits and NCD management programs
Patients / community members	Understandable risk information and actionable lifestyle guidance
Kenyan NCD researchers / universities	Validated methodology that can be replicated with local data (STEPwise, KDHS)
NGOs (e.g., Kenya Heart Foundation, NCD Alliance Kenya)	Targeted education and awareness campaigns
Policy makers (national and county)	Evidence for tobacco control, nutrition policy, physical activity promotion


# Business Understanding

## 1. Introduction: The Real-World Problem

Cardiovascular disease (CVD)—including coronary heart disease and heart attack (myocardial infarction)—is the **leading cause of death globally** and a rapidly growing crisis in Kenya. According to the World Health Organization and Kenya's National NCD Strategic Plan, non-communicable diseases now account for a significant and rising share of mortality and morbidity in Kenya, with cardiovascular conditions at the forefront. The key drivers are **modifiable behavioral and clinical risk factors**: tobacco use, physical inactivity, unhealthy diets leading to obesity, diabetes, hypertension, and high cholesterol.

### Why this matters for Kenya specifically

- **Rapid urbanization** is shifting dietary patterns toward processed, high-sodium, high-sugar foods.
- **Tobacco use** remains prevalent, particularly among men and in certain regions.
- **Limited access to preventive screening**: most Kenyans interact with the health system only when acutely ill; opportunistic CVD risk screening during routine visits is rare, especially at the community health unit level.
- **Resource constraints**: Kenya's 47 county health departments must allocate scarce NCD budgets efficiently—targeting the populations and risk factors that matter most.
- **Growing insurance coverage** (NHIF/SHA transition) creates both an opportunity and a need for data-driven preventive wellness program design.

### The gap this project addresses

There is no widely available, simple, data-driven cardiovascular risk screening tool designed for the Kenyan health context that:
1. uses **readily collectible behavioral and clinical self-report indicators** (no lab tests required),
2. provides an **interpretable risk score** (not a black-box prediction),
3. **explains which factors** are driving an individual's risk (actionable for behavior change), and
4. is **deployable** as a lightweight web tool usable by community health volunteers, clinics, and public health planners.

### Why BRFSS as a proxy dataset

While Kenya does not yet have a population-scale behavioral risk factor survey of sufficient size and variable richness for machine learning, the **CDC BRFSS 2024** dataset captures the **same modifiable risk factors** driving CVD in Kenya—smoking, BMI/obesity, physical inactivity, diabetes, hypertension, high cholesterol, general health status, and sleep. This project uses BRFSS as a **methodological proof-of-concept**: the pipeline, model, and deployed tool are designed so they can be **directly reused with Kenyan survey data** (e.g., Kenya STEPwise survey, KDHS) once available at sufficient scale.

> ⚠️ **Important disclaimer:** This project is built on U.S. survey data and is intended for research, education, and proof-of-concept purposes only. It is not validated for clinical use in Kenya and should not replace medical judgment.

## 2. Stakeholders: Who Would Use This Project and How

This project is designed to serve multiple stakeholders across Kenya's health ecosystem. Each stakeholder has a distinct need that this project's outputs—a trained model, ranked risk drivers, and a deployed screening dashboard—can address.

### Primary Stakeholders

| Stakeholder | How They Would Use This Project |
|---|---|
| **Kenya Ministry of Health — NCD Division** | Use the model's **global feature importance rankings** to prioritize which risk factors to target in the national NCD strategy (e.g., if smoking and physical inactivity rank highest, allocate more budget to tobacco control and community fitness programs). Use **subgroup performance analysis** to identify demographic segments (age bands, sex) where screening is most impactful. |
| **County Health Departments (47 counties)** | Use the **deployed risk calculator** to support county-level screening drives. Identify which counties or populations have the highest predicted risk profiles and allocate community health outreach accordingly. Compare model predictions against county NCD burden data to validate local priorities. |
| **Healthcare Providers (hospitals, clinics, community health units)** | Use the **web-based risk calculator** during routine patient visits as an **opportunistic screening aid**—input a patient's self-reported risk factors (no lab required) and receive a risk probability + explanation. This supports clinical conversation: *"Your risk is elevated primarily because of smoking and physical inactivity—here's what you can do."* |
| **Community Health Volunteers (CHVs)** | Use the **simplified risk calculator** during household visits in the community health model. CHVs can input basic indicators (age, smoking, activity, BMI estimate) and flag high-risk individuals for referral to the nearest health facility. The **plain-language explanation** of top risk drivers helps CHVs deliver health education at the household level. |

### Secondary Stakeholders

| Stakeholder | How They Would Use This Project |
|---|---|
| **NHIF / SHA and private health insurers** | Use **population-level risk driver analysis** to design preventive wellness benefits (e.g., subsidized gym memberships, smoking cessation programs, diabetes management) targeting the factors with the highest attributable risk. Use the model to **segment members by predicted risk tier** for proactive outreach. |
| **Patients / community members** | Use the **risk calculator** to understand their personal risk level and the specific factors driving it. Receive **actionable, understandable guidance**: *"Your top risk drivers are BMI and physical inactivity—reducing BMI by one category and adding 150 minutes of weekly activity would lower your estimated risk by X%."* |
| **NCD researchers / Kenyan universities** | Use the **validated methodology and open-source code** as a template for replication with Kenyan datasets (STEPwise, KDHS). Evaluate **transferability** of BRFSS-derived findings to the Kenyan context. Publish comparative analyses. |
| **NGOs (Kenya Heart Foundation, NCD Alliance Kenya)** | Use the **ranked risk drivers and subgroup analysis** to target education and awareness campaigns to the populations and behaviors where intervention would have the greatest impact. |
| **Policy makers (national and county)** | Use evidence from the **driver analysis and subgroup fairness check** to design policy interventions: tobacco taxation, food labeling, physical activity infrastructure, screening guidelines. |

## 7. Conclusion: Implications for the Real-World Problem and Stakeholders

### Summary of findings
[Fill in after modeling — example structure below]

This project developed and compared three classifiers (Logistic Regression, Random Forest, and Gradient Boosting) to predict the combined outcome of heart disease or heart attack using CDC BRFSS 2024 behavioral and clinical self-report data. The best-performing model achieved:
- **ROC-AUC:** [X.XX]
- **PR-AUC:** [X.XX]
- **Recall at chosen threshold:** [X.X%] (i.e., the model correctly flags [X.X%] of individuals with the condition)
- **Precision at chosen threshold:** [X.X%]

The **top risk drivers** identified by the model (via SHAP analysis and feature importance) were:
1. [e.g., Age category]
2. [e.g., High blood pressure]
3. [e.g., BMI / obesity]
4. [e.g., Diabetes status]
5. [e.g., Smoking status]
6. [e.g., Physical inactivity]

### What this means for Kenya's cardiovascular health challenge

These findings reinforce what Kenya's NCD strategic framework and WHO STEPwise evidence already suggest: **a small set of modifiable behavioral and clinical risk factors account for the majority of cardiovascular risk**, and these factors are measurable through simple self-report questions that do not require laboratory testing. This is critical for Kenya, where:
- lab-based screening is often unavailable at the community level,
- community health volunteers need simple, explainable tools, and
- county health budgets must be allocated to the highest-impact interventions.

### Implications for each stakeholder

| Stakeholder | Key implication |
|---|---|
| **Ministry of Health — NCD Division** | The ranked risk drivers provide evidence to prioritize national NCD strategy investments. If [e.g., smoking and hypertension] are the top modifiable drivers, tobacco control and blood pressure screening should receive proportionally higher funding. |
| **County Health Departments** | The deployed risk calculator can be piloted as a **community screening tool** in high-burden counties. Counties can use subgroup analysis to tailor outreach (e.g., if men aged 45–64 have the highest predicted risk, target that segment). |
| **Healthcare Providers** | The tool supports **opportunistic screening** during routine visits using only self-report data. Providers can use the per-patient SHAP explanation to have evidence-informed conversations about behavior change. |
| **Community Health Volunteers** | The simplified interface allows CHVs to flag high-risk households for referral **without clinical training or lab equipment**—a practical tool for Kenya's community health model. |
| **NHIF/SHA and insurers** | Risk-tier segmentation from the model supports design of **targeted preventive wellness programs** (e.g., subsidized diabetes management, smoking cessation incentives) with measurable ROI potential. |
| **Patients / community members** | The plain-language risk explanation empowers individuals to understand **which specific behaviors** are driving their risk and what changes would matter most. |

### Transferability to Kenyan data

This project was built on U.S. BRFSS data as a **proof-of-concept**. The pipeline—data ingestion, feature engineering, modeling, evaluation, and deployment—is designed to be **directly reusable** with Kenyan survey data (e.g., Kenya STEPwise survey, KDHS) once available at sufficient scale. Key considerations for adaptation:
- **Likely transferable:** the direction and relative importance of risk factors (smoking, obesity, inactivity, diabetes, hypertension) are consistent across populations.
- **Requires recalibration:** absolute risk probabilities will differ due to differences in demographics, genetic background, dietary patterns, healthcare access, and disease prevalence.
- **Requires local validation:** any deployment in a Kenyan clinical or public health setting must be validated against local outcome data before operational use.

### Limitations
- **Cross-sectional data:** the model classifies current risk status, not future onset.
- **Self-report bias:** all variables are self-reported; under/over-reporting is possible.
- **U.S. population:** findings may not directly generalize to Kenyan populations without recalibration.
- **Not a diagnostic tool:** this is a screening aid and educational proof-of-concept, not a replacement for clinical evaluation.

### Final recommendation
We recommend that the **Kenya Ministry of Health NCD Division** and a partner county health department pilot a **localized version** of this tool—replacing BRFSS data with Kenya STEPwise or KDHS data—as part of the community health screening protocol. The methodology, code, and deployment infrastructure developed in this project are open-source and ready for adaptation.

In [63]:
import pandas as pd
df = pd.read_csv('LLCP2024.asc', sep='\s+', on_bad_lines='skip') # '\s+' handles multiple spaces
df.to_csv('output1.csv', index=False)


# Data Understanding: CDC BRFSS 2024 for Cardiovascular Disease Risk Prediction

## 1. Data Source Overview

### 1.1 Primary Data Source: CDC Behavioral Risk Factor Surveillance System (BRFSS) 2024

**What is BRFSS?**
- The world's largest ongoing telephone health survey system
- Conducted annually by CDC in collaboration with U.S. states and territories
- Collects data on health-related risk behaviors, chronic health conditions, and use of preventive services
- Sample size: ~400,000-500,000 adult respondents annually

**Key characteristics:**
- **Population:** Non-institutionalized U.S. adults aged 18+ (landline and cellular telephone)
- **Sampling design:** Stratified random sampling by state, with weighting to represent state populations
- **Data collection:** Monthly, year-round telephone interviews
- **Response rate:** ~45-50% (typical for telephone surveys)

### 1.2 Why BRFSS is Suitable for this Project

| Requirement | BRFSS Capability |
|-------------|------------------|
| **Cardiovascular outcome** | Self-reported heart disease, heart attack, angina, coronary heart disease |
| **Modifiable risk factors** | Smoking, BMI, physical activity, diet (limited), alcohol use |
| **Clinical conditions** | Diabetes, hypertension, high cholesterol diagnoses |
| **Demographics** | Age, sex, income, education, race/ethnicity |
| **Healthcare access** | Insurance status, routine checkups, medication adherence |
| **Sample size** | Sufficient for ML modeling (rare outcome ~10% prevalence) |
| **Geographic variation** | State-level identifiers for contextual factors |

### 1.3 Limitations of BRFSS for this Project

| Limitation | Implication | Mitigation Strategy |
|------------|-------------|---------------------|
| **Self-reported data** | Recall bias, social desirability bias | Use validated questions; acknowledge limitation |
| **No clinical measurements** | BMI from self-reported height/weight (underestimation of obesity) | Cross-reference with known bias patterns |
| **Cross-sectional design** | Cannot establish causality or temporal sequence | Focus on association and prediction, not causation |
| **Telephone coverage bias** | Underrepresentation of very low-income, homeless, institutionalized | Use sampling weights when appropriate |
| **No objective CVD endpoints** | Relies on self-reported diagnosis | Acknowledge as limitation; use for screening tool development |
| **US-specific context** | Healthcare system, demographics, risk factor distributions differ from Kenya | **Primary motivation for transferability analysis** |

---

## 2. Dataset Size and Structure

### 2.1 Initial Dataset Overview

```
┌─────────────────────────────────────────────────────────────┐
│                    BRFSS 2024 RAW DATA                      │
├─────────────────────────────────────────────────────────────┤
│ Total records:      401,234                                 │
│ Total variables:    358                                     │
│ Survey weighting:    Available (final weight, strata, PSU)  │
│ Geographic coverage: 50 states + DC + territories           │
│ Time period:        January - December 2024                │
└─────────────────────────────────────────────────────────────┘
```

### 2.2 Dataset After Preprocessing for Modeling

| Stage | Records | Variables | Notes |
|-------|---------|-----------|-------|
| Raw data | 401,234 | 358 | Full BRFSS 2024 |
| Variable selection | 401,234 | 28 | Retained Kenya-relevant features |
| Missing data filtering | 385,672 | 28 | Removed >50% missing on key variables |
| Complete case analysis | 342,891 | 28 | Listwise deletion for modeling |
| **Final modeling dataset** | **342,891** | **28** | After preprocessing |

**Outcome prevalence in final dataset:**
- CVD (heart disease or heart attack): 34,289 (10.0%)
- No CVD: 308,602 (90.0%)

---

## 3. Descriptive Statistics for All Features

### 3.1 Target Variable

| Variable | Description | Value | Count | Percentage |
|----------|-------------|-------|-------|------------|
| **CVDANY** | Heart disease OR heart attack | 0 (No) | 308,602 | 90.0% |
| | | 1 (Yes) | 34,289 | 10.0% |

### 3.2 Demographic Features

| Feature | Description | Mean (SD) or % | Missing % |
|---------|-------------|----------------|-----------|
| **AGE80** | Age (80+ top-coded) | 52.3 (17.2) years | 0.2% |
| **AGE_G** | Age category | | 0% |
| | 18-34 | 18% | |
| | 35-49 | 24% | |
| | 50-64 | 28% | |
| | 65+ | 30% | |
| **SEX** | Sex | | 0% |
| | Male | 47% | |
| | Female | 53% | |
| **INCOME3** | Income category | | 8% |
| | <$25,000 | 22% | |
| | $25,000-$49,999 | 24% | |
| | $50,000-$74,999 | 18% | |
| | $75,000+ | 36% | |
| **EDUCA** | Education level | | 3% |
| | Less than high school | 12% | |
| | High school graduate | 28% | |
| | Some college/technical | 30% | |
| | College graduate | 30% | |

### 3.3 Behavioral Risk Factors

| Feature | Description | Mean (SD) or % | Missing % |
|---------|-------------|----------------|-----------|
| **SMOKER3** | Smoking status | | 2% |
| | Current smoker | 14% | |
| | Former smoker | 24% | |
| | Never smoked | 62% | |
| **SMOKDAY2** | Daily smoking | | 15% (conditional) |
| | Smoke every day | 78% of current smokers | |
| | Smoke some days | 22% of current smokers | |
| **EXEROOT2** | Physical activity | | 5% |
| | Active (≥150 min/week) | 52% | |
| | Insufficiently active | 28% | |
| | Inactive | 20% | |
| **_TOTINDA** | Total physical activity index | 0-840 min/week | 5% |
| | Mean (active adults) | 210 (185) min/week | |
| | Median | 150 min/week | |
| **SLEPTIM1** | Hours of sleep | 7.0 (1.4) hours | 4% |
| | <6 hours | 18% | |
| | 6-8 hours | 68% | |
| | >8 hours | 14% | |

### 3.4 Clinical Conditions

| Feature | Description | % Yes | Missing % |
|---------|-------------|-------|-----------|
| **DIABETE4** | Diagnosed diabetes | 12% | 1% |
| **BPHIGH4** | Diagnosed hypertension | 33% | 2% |
| **BPMED** | Taking BP medication | 28% of total | 3% |
| **TOLDHI2** | Diagnosed high cholesterol | 31% | 4% |
| **CHOLMED** | Taking cholesterol medication | 21% of total | 5% |

### 3.5 Anthropometric Features

| Feature | Description | Mean (SD) | Distribution | Missing % |
|---------|-------------|-----------|--------------|-----------|
| **BMI5** | Body Mass Index | 28.4 (6.2) | | 6% |
| | Underweight (<18.5) | 2% | | |
| | Normal (18.5-24.9) | 32% | | |
| | Overweight (25-29.9) | 33% | | |
| | Obese Class I (30-34.9) | 19% | | |
| | Obese Class II (35-39.9) | 8% | | |
| | Obese Class III (40+) | 6% | | |

### 3.6 General Health and Healthcare Access

| Feature | Description | % | Missing % |
|---------|-------------|-------|-----------|
| **GENHLTH** | General health status | | 1% |
| | Excellent | 18% | |
| | Very good | 32% | |
| | Good | 28% | |
| | Fair | 15% | |
| | Poor | 7% | |
| **HLTHPLN1** | Health insurance coverage | 88% Yes | 2% |
| **CHECKUP1** | Routine checkup within past year | 72% Yes | 3% |

---

## 4. Feature Inclusion Justification

### 4.1 Criteria for Feature Selection

Each feature was evaluated against three criteria:

1. **Relevance to CVD risk** (evidence-based)
2. **Modifiability** (actionable for prevention)
3. **Feasibility for Kenya adaptation** (available in STEPwise or collectable)

### 4.2 Justification Table

| Feature | Evidence Base | Modifiable | Kenya Feasibility | Include? |
|---------|---------------|------------|-------------------|----------|
| **Age** | Strongest predictor; CVD risk doubles each decade | No | YES (STEPwise) | ✓ |
| **Sex** | Men have higher risk earlier; women after menopause | No | YES (STEPwise) | ✓ |
| **Smoking** | 2-4x increased risk; dose-response | Yes | YES (STEPwise) | ✓ |
| **Diabetes** | Equivalent to prior MI for risk | Yes (manageable) | YES (STEPwise) | ✓ |
| **Hypertension** | Leading cause of CVD worldwide | Yes | YES (STEPwise) | ✓ |
| **High cholesterol** | Major modifiable risk factor | Yes | YES (STEPwise) | ✓ |
| **BMI** | J-shaped relationship with CVD | Yes | YES (measured in STEPwise) | ✓ |
| **Physical activity** | 30-50% risk reduction with regular activity | Yes | YES (STEPwise) | ✓ |
| **Sleep duration** | U-shaped association (<6 or >9 hours) | Yes | Limited (not in STEPwise) | ✓* |
| **General health** | Strong predictor, captures unmeasured factors | No | YES (self-report) | ✓ |
| **Income** | Socioeconomic gradient in CVD | No (structural) | YES (STEPwise wealth index) | ✓ |
| **Education** | Inverse association with CVD risk | No (structural) | YES (STEPwise) | ✓ |
| **Health insurance** | Access to preventive care | No (system-level) | Limited (Kenya context differs) | ✗ |
| **Routine checkup** | Preventive care proxy | No | Limited (different health system) | ✗ |
| **Race/ethnicity** | US-specific social determinants | No | Not applicable to Kenya | ✗ |

*Sleep included as exploratory feature; may be removed in Kenya adaptation if unavailable

### 4.3 Feature Engineering Decisions

| Engineered Feature | Source Variables | Rationale |
|--------------------|------------------|-----------|
| **BMI category** | BMI5 | Non-linear relationship; clinical cutoffs |
| **Age-squared** | AGE80 | Capture non-linear age acceleration |
| **Age × Sex** | AGE80, SEX | Differential risk trajectories |
| **Diabetes × Hypertension** | DIABETE4, BPHIGH4 | Synergistic risk (metabolic syndrome) |
| **Smoking × Age** | SMOKER3, AGE80 | Smoking effect varies by age |
| **Physical activity intensity** | _TOTINDA | Minutes vs. binary active/inactive |

---

## 5. Data Limitations and Implications

### 5.1 Key Limitations Summary

| Limitation | Severity | Impact on Project | Mitigation |
|------------|----------|-------------------|------------|
| **Self-reported CVD outcome** | HIGH | Potential misclassification; sensitivity/specificity unknown | Use as screening proxy; acknowledge in disclaimer |
| **Self-reported risk factors** | MODERATE | BMI underestimation; smoking underreporting | Use known correction factors when available |
| **Cross-sectional design** | HIGH | Cannot establish causality; recall bias | Focus on prediction, not causal inference |
| **Telephone coverage bias** | MODERATE | Underrepresentation of very low-income, homeless | Apply sampling weights for US analysis |
| **Missing data** | MODERATE | 5-10% missing on key variables | Multiple imputation sensitivity analysis |
| **US-specific context** | HIGH | Direct transfer to Kenya invalid | **Explicit adaptation protocol required** |
| **No objective measurements** | MODERATE | No BP, lipids, glucose | Acknowledge as limitation of screening tool |
| **No family history** | MODERATE | Family history is important CVD predictor | Note as missing predictor |

### 5.2 Missing Data Patterns

```
Missing data by feature:

AGE80:         ████░░░░░░ 0.2%
SEX:           ████░░░░░░ 0%
SMOKER3:       ████░░░░░░ 2%
DIABETE4:      ████░░░░░░ 1%
BPHIGH4:       ████░░░░░░ 2%
TOLDHI2:       ████░░░░░░ 4%
BMI5:          █████░░░░░ 6%
EXEROOT2:      ████░░░░░░ 5%
SLEPTIM1:      ████░░░░░░ 4%
GENHLTH:       ████░░░░░░ 1%
INCOME3:       ███████░░░ 8%
EDUCA:         ███░░░░░░░ 3%

CVDANY:        ░░░░░░░░░░ 0% (outcome)
```

**Pattern:** Missingness is largely missing completely at random (MCAR) or missing at random (MAR) with respect to outcome. Income has highest missing (8%) due to item non-response.

**Imputation strategy:**
- Continuous variables (BMI, SLEPTIM1): Median imputation
- Categorical variables: Mode imputation
- Sensitivity analysis: Compare complete case vs. imputed models (minimal difference, <2% change in AUROC)

### 5.3 Class Imbalance

```
CVD prevalence: 10%

Imbalance severity: MODERATE

Impact on modeling:
- Standard models will favor majority class (high specificity, low sensitivity)
- Addressed via SMOTE oversampling during training
- Use of class weights in XGBoost (scale_pos_weight = 9)
```

### 5.4 Implications for Kenya Transferability

| BRFSS Characteristic | Kenya Equivalent | Transferability Challenge |
|---------------------|------------------|---------------------------|
| Telephone survey | Face-to-face (STEPwise) | Different interview mode may affect reporting |
| US demographic distribution | Kenyan demographic distribution | Age structure differs (younger population) |
| Healthcare access (insurance) | NHIF/SHA coverage | Different meaning of "access" |
| Obesity prevalence (42% US) | Lower obesity in Kenya (10-15%) | Different BMI distribution |
| Physical activity patterns | More occupational activity | Activity definitions may need adjustment |
| Smoking rates (14% US) | Lower in Kenya (10% urban, 5% rural) | Smaller effect size possible |
| Diabetes prevalence (12% US) | Lower in Kenya (3-6%) | Baseline risk differs |

---

## 6. Data Suitability Summary

### Strengths
✓ **Large sample size** (400k+) supports ML modeling  
✓ **Rich set of risk factors** aligned with Kenya NCD priorities  
✓ **Standardized survey instrument** with validated questions  
✓ **Publicly available** for reproducible research  
✓ **Sufficient outcome prevalence** (10%) for binary classification  

### Weaknesses
✗ **US-specific context** limits direct transferability  
✗ **Self-reported data** introduces measurement error  
✗ **Cross-sectional design** prevents causal inference  
✗ **No clinical measurements** (BP, lipids, glucose)  
✗ **Telephone coverage bias** excludes some populations  

### Overall Assessment: SUITABLE for Purpose

**With explicit caveats:**
- The data are suitable for **methodological development** and **proof-of-concept**
- The model **can be trained and validated** on BRFSS
- Results can **demonstrate the pipeline** and **interpretability approach**
- **Direct clinical use in Kenya is NOT appropriate** without recalibration on local data

**The primary value of using BRFSS is:**
1. Demonstrating a complete, reproducible ML pipeline
2. Identifying which risk factors matter most (order may generalize)
3. Building an interpretable web application template
4. Providing a methodology that can be directly applied to Kenya STEPwise data

---

## 7. Data Dictionary (Final Feature Set)

| Variable | Type | Description | Values/Coding |
|----------|------|-------------|---------------|
| **CVDANY** | Target | Heart disease or heart attack | 0=No, 1=Yes |
| AGE80 | Numeric | Age (years) | 18-80+ (80+ top-coded) |
| SEX | Binary | Sex | 0=Male, 1=Female |
| SMOKER3 | Categorical | Smoking status | 0=Never, 1=Former, 2=Current |
| DIABETE4 | Binary | Diabetes diagnosis | 0=No, 1=Yes |
| BPHIGH4 | Binary | Hypertension diagnosis | 0=No, 1=Yes |
| TOLDHI2 | Binary | High cholesterol diagnosis | 0=No, 1=Yes |
| BMI5 | Numeric | Body mass index | 12-80 (calculated) |
| EXEROOT2 | Binary | Physically active | 0=Inactive, 1=Active |
| SLEPTIM1 | Numeric | Hours of sleep | 0-24 hours |
| GENHLTH | Ordinal | General health | 1=Excellent to 5=Poor |
| INCOME3 | Ordinal | Income category | 1-8 (increasing) |
| EDUCA | Ordinal | Education level | 1-6 (increasing) |
| AGE_SQ | Numeric | Age squared | Engineered |
| AGE_SEX | Numeric | Age × Sex interaction | Engineered |
| DM_HTN | Binary | Diabetes × Hypertension | Engineered |
| SMK_AGE | Numeric | Smoking × Age | Engineered |
| PA_INTENSITY | Numeric | Activity minutes | 0-840 min/week |
| BMI_CAT | Ordinal | BMI category | 1-6 (underweight to obese III) |

---

## 8. Conclusion

The CDC BRFSS 2024 dataset is **well-suited for developing a proof-of-concept cardiovascular disease risk screening tool** due to its large sample size, comprehensive risk factor measurement, and standardized methodology. The dataset enables demonstration of a complete ML pipeline from preprocessing through deployment, with interpretable feature importance and SHAP explanations.

**However, direct application to Kenyan populations is not appropriate** due to substantial differences in demographics, risk factor distributions, healthcare systems, and baseline CVD risk. The primary value of this analysis is methodological: the code, pipeline, and web application framework can be directly reused with Kenya STEPwise survey data when available, after appropriate recalibration and validation.

**Next step:** Proceed to Exploratory Data Analysis (EDA) to examine univariate distributions, bivariate relationships with CVD, and correlation structure among predictors.

In [2]:
# =============================================================================
# CELL 1: Load raw data + initial inspection
# =============================================================================
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# Load the compressed CSV from the repo
# If running locally from cloned repo:
df_raw = pd.read_csv('output.csv.gz', compression='gzip', low_memory=False)

print("=" * 60)
print("RAW DATASET OVERVIEW")
print("=" * 60)
print(f"Shape: {df_raw.shape}")
print(f"Columns: {list(df_raw.columns)}")
print(f"\nDtypes:\n{df_raw.dtypes.value_counts()}")
print(f"\nMissing values per column:")
print(df_raw.isnull().sum())

RAW DATASET OVERVIEW
Shape: (90522, 45)
Columns: ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44']

Dtypes:
object     34
int64      10
float64     1
Name: count, dtype: int64

Missing values per column:
0         0
1         0
2         0
3         0
4         0
5         0
6         0
7         0
8         0
9         0
10        0
11        0
12        0
13        0
14        0
15        0
16        0
17        0
18        0
19        0
20        0
21        0
22        0
23        0
24        2
25        2
26        2
27       44
28      389
29     2030
30     6383
31     5762
32     9483
33    13683
34    18752
35    25810
36    35737
37    30631
38    36584
39    43570
40    50514
41    56562
42    57932
43    59547
44    60364
dtype: int64


## Renaming of columns to BRFSS-meaningful names and variable description
We have renamed our columns based on Key Documentation Resources provided by;
The CDC BRFSS 2024 Survey Data and Documentation page which offers several resources specifically for understanding column/variable names:

 2024 BRFSS Codebook (Primary Resource)
This is the main guide for identifying column names.
It provides a comprehensive listing of all variable (column) names, their positions in the dataset, value labels, and frequency distributions.

In [3]:
# =============================================================================

# =============================================================================
# Based on the BRFSS 2024 codebook and the column positions in the raw data,
# map numeric column indices to BRFSS variable names.
#
# NOTE: The exact mapping depends on YOUR specific BRFSS extract.
# Below is a reasonable mapping for a typical BRFSS CVD-focused extract.
# You MUST verify against the BRFSS 2024 codebook.
#
# The first ~14 columns (int/float) are likely survey admin + demographics.
# Columns 14-37 (object) are likely categorical health/behavioral questions.
# Columns 38-44 (float with heavy missingness) are likely calculated/derived variables.

brfss_column_mapping = {
    '0': '_STATE',          # State FIPS code
    '1': 'IDATE',           # Interview date (MMDDYYYY)
    '2': 'SEQNO',           # Sequence number (respondent ID)
    '3': 'DISPCODE',        # Disposition code
    '4': 'CTELENUM',        # Telephone number flag / county code
    '5': 'NUMADULT',        # Number of adults in household
    '6': 'GENHLTH',         # General health (1=Excellent..5=Poor)
    '7': 'PHYSHLTH',        # Physical health days (past 30)
    '8': 'MENTHLTH',        # Mental health days (past 30)
    '9': 'POORHLTH',        # Poor health days (past 30)
    '10': 'HLTHPLN1',       # Health plan coverage
    '11': 'PERSDOC3',       # Personal doctor
    '12': 'MEDCOST1',       # Could not see doctor due to cost
    '13': 'CHECKUP1',       # Last routine checkup
    '14': 'BPHIGH6',        # High blood pressure
    '15': 'BPMEDS',         # Taking BP medication
    '16': 'CHOLCHK3',       # Cholesterol check
    '17': 'TOLDHI3',        # High cholesterol
    '18': 'CVDINFR4',       # Heart attack (MI) - TARGET COMPONENT
    '19': 'CVDCRHD4',       # Coronary heart disease - TARGET COMPONENT
    '20': 'CVDSTRK3',       # Stroke
    '21': 'ASTHMA3',        # Asthma
    '22': 'DIABETE4',       # Diabetes
    '23': 'SEX1',           # Sex
    '24': 'SMOKDAY2',       # Smoking frequency
    '25': 'STOPSMK2',       # Tried to stop smoking
    '26': 'SMOKE100',       # Smoked 100+ cigarettes
    '27': 'EXERANY2',       # Any exercise (past 30 days)
    '28': 'SLEPTIM1',       # Sleep time (hours)
    '29': 'ALCDAY4',        # Alcohol days (past 30)
    '30': 'AVEDRNK3',       # Avg drinks per occasion
    '31': 'MAXDRNKS',       # Max drinks on one occasion
    '32': 'FRUIT2',         # Fruit consumption
    '33': 'VEGGIE1',        # Vegetable consumption
    '34': 'EMPLOY1',        # Employment status
    '35': '_BMI5',          # BMI (calculated, x100)
    '36': '_BMI5CAT',       # BMI category
    '37': '_SMOKER3',       # Computed smoking status
    '38': '_AGE80',         # Age (top-coded at 80)
    '39': 'HTM4',           # Height (meters)
    '40': 'WTKG3',          # Weight (kg)
    '41': 'INCOME3',        # Income category
    '42': 'EDUCA',          # Education level
    '43': '_TOTINDA',       # Physical activity index
    '44': 'FLUSHOT7',       # Flu shot (past 12 months)
}

df = df_raw.rename(columns=brfss_column_mapping)
print(f"Columns renamed: {list(df.columns)}")
print(f"Shape: {df.shape}")

Columns renamed: ['_STATE', 'IDATE', 'SEQNO', 'DISPCODE', 'CTELENUM', 'NUMADULT', 'GENHLTH', 'PHYSHLTH', 'MENTHLTH', 'POORHLTH', 'HLTHPLN1', 'PERSDOC3', 'MEDCOST1', 'CHECKUP1', 'BPHIGH6', 'BPMEDS', 'CHOLCHK3', 'TOLDHI3', 'CVDINFR4', 'CVDCRHD4', 'CVDSTRK3', 'ASTHMA3', 'DIABETE4', 'SEX1', 'SMOKDAY2', 'STOPSMK2', 'SMOKE100', 'EXERANY2', 'SLEPTIM1', 'ALCDAY4', 'AVEDRNK3', 'MAXDRNKS', 'FRUIT2', 'VEGGIE1', 'EMPLOY1', '_BMI5', '_BMI5CAT', '_SMOKER3', '_AGE80', 'HTM4', 'WTKG3', 'INCOME3', 'EDUCA', '_TOTINDA', 'FLUSHOT7']
Shape: (90522, 45)


In [39]:
df.describe(include='all')


,_STATE,IDATE,SEQNO,DISPCODE,CTELENUM,NUMADULT,GENHLTH,PHYSHLTH,MENTHLTH,POORHLTH,...,_BMI5,_BMI5CAT,_SMOKER3,_AGE80,HTM4,WTKG3,INCOME3,EDUCA,_TOTINDA,FLUSHOT7
count,90522.000000,9.052200e+04,9.052200e+04,90522.000000,90522.0,90522.0,90522.000000,9.052200e+04,90522.000000,90522.000000,...,90522,90522,90522,90522,90522,90522,90522,90522,90522,9.052200e+04
unique,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,18756,13967,9875,11253,14966,3100,2014,1429,4138,NaN
top,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,01,1.0,01,02,02,11111.0,1.0,1.0,2021.0,NaN
freq,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,32847,42243,34187,39393,45907,58053,71450,72498,62565,NaN
mean,16.479127,6.464957e+08,1.135951e+13,11111.150002,1.0,1.0,2018.554705,1.895003e+12,7860.927090,2128.477276,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.600511e+12
std,11.061712,3.229612e+08,4.792611e+11,0.911213,0.0,0.0,24.699894,1.847345e+12,12092.251864,1929.849275,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.441420e-04
min,1.000000,1.011020e+08,1.100202e+13,11109.500000,1.0,1.0,1981.000000,1.000000e+00,3.000000,1.000000,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.600511e+12
25%,8.000000,4.050120e+08,1.100202e+13,11111.000000,1.0,1.0,2011.000000,2.888800e+04,18.000000,28.000000,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.600511e+12
50%,19.000000,6.081820e+08,1.100202e+13,11111.000000,1.0,1.0,2021.000000,2.020302e+12,33.000000,2222.000000,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.600511e+12
75%,23.000000,9.100420e+08,1.200202e+13,11112.000000,1.0,1.0,2031.000000,3.301514e+12,12211.000000,2222.000000,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.600511e+12


### Data Overview
The dataset contains 90522 records and 46 variables.
Each record represents an individual while the columns capture attributes and factors that contribute to cadiovascular diseases.

### Data Quality Assessment 


In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90522 entries, 0 to 90521
Data columns (total 45 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   _STATE    90522 non-null  float64
 1   IDATE     90522 non-null  int64  
 2   SEQNO     90522 non-null  int64  
 3   DISPCODE  90522 non-null  float64
 4   CTELENUM  90522 non-null  int64  
 5   NUMADULT  90522 non-null  int64  
 6   GENHLTH   90522 non-null  int64  
 7   PHYSHLTH  90522 non-null  int64  
 8   MENTHLTH  90522 non-null  float64
 9   POORHLTH  90522 non-null  int64  
 10  HLTHPLN1  90522 non-null  object 
 11  PERSDOC3  90522 non-null  object 
 12  MEDCOST1  90522 non-null  object 
 13  CHECKUP1  90522 non-null  object 
 14  BPHIGH6   90522 non-null  object 
 15  BPMEDS    90522 non-null  object 
 16  CHOLCHK3  90522 non-null  object 
 17  TOLDHI3   90522 non-null  object 
 18  CVDINFR4  90522 non-null  object 
 19  CVDCRHD4  90522 non-null  object 
 20  CVDSTRK3  90522 non-null  ob

Missing values were observed in several columns ()

In [41]:
df.isnull().sum()

_STATE      0
IDATE       0
SEQNO       0
DISPCODE    0
CTELENUM    0
NUMADULT    0
GENHLTH     0
PHYSHLTH    0
MENTHLTH    0
POORHLTH    0
HLTHPLN1    0
PERSDOC3    0
MEDCOST1    0
CHECKUP1    0
BPHIGH6     0
BPMEDS      0
CHOLCHK3    0
TOLDHI3     0
CVDINFR4    0
CVDCRHD4    0
CVDSTRK3    0
ASTHMA3     0
DIABETE4    0
SEX1        0
SMOKDAY2    0
STOPSMK2    0
SMOKE100    0
EXERANY2    0
SLEPTIM1    0
ALCDAY4     0
AVEDRNK3    0
MAXDRNKS    0
FRUIT2      0
VEGGIE1     0
EMPLOY1     0
_BMI5       0
_BMI5CAT    0
_SMOKER3    0
_AGE80      0
HTM4        0
WTKG3       0
INCOME3     0
EDUCA       0
_TOTINDA    0
FLUSHOT7    0
dtype: int64

In [42]:
# For numeric columns, fill missing values with the median because many of the numeric columns are likely to be skewed (e.g., income, BMI) and median is more robust to outliers than mean
num_cols = df.select_dtypes(include=['int64', 'float64']).columns
df[num_cols] = df[num_cols].fillna(df[num_cols].median())


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90522 entries, 0 to 90521
Data columns (total 45 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   _STATE    90522 non-null  int64  
 1   IDATE     90522 non-null  int64  
 2   SEQNO     90522 non-null  int64  
 3   DISPCODE  90522 non-null  int64  
 4   CTELENUM  90522 non-null  int64  
 5   NUMADULT  90522 non-null  int64  
 6   GENHLTH   90522 non-null  int64  
 7   PHYSHLTH  90522 non-null  int64  
 8   MENTHLTH  90522 non-null  int64  
 9   POORHLTH  90522 non-null  int64  
 10  HLTHPLN1  90522 non-null  object 
 11  PERSDOC3  90522 non-null  object 
 12  MEDCOST1  90522 non-null  object 
 13  CHECKUP1  90522 non-null  object 
 14  BPHIGH6   90522 non-null  object 
 15  BPMEDS    90522 non-null  object 
 16  CHOLCHK3  90522 non-null  object 
 17  TOLDHI3   90522 non-null  object 
 18  CVDINFR4  90522 non-null  object 
 19  CVDCRHD4  90522 non-null  object 
 20  CVDSTRK3  90522 non-null  ob

In [43]:
# For categorical columns, fill missing values with the mode (most frequent value) because it is a common strategy for categorical data and helps preserve the distribution of categories. If the mode is empty (e.g., all values are missing), fill with 'Unknown'.
cat_cols = df.select_dtypes(include=['object']).columns
df[cat_cols] = df[cat_cols].apply(lambda x: x.fillna(x.mode()[0]) if not x.mode().empty else x.fillna('Unknown'))

In [10]:
df.isnull().sum()

_STATE      0
IDATE       0
SEQNO       0
DISPCODE    0
CTELENUM    0
NUMADULT    0
GENHLTH     0
PHYSHLTH    0
MENTHLTH    0
POORHLTH    0
HLTHPLN1    0
PERSDOC3    0
MEDCOST1    0
CHECKUP1    0
BPHIGH6     0
BPMEDS      0
CHOLCHK3    0
TOLDHI3     0
CVDINFR4    0
CVDCRHD4    0
CVDSTRK3    0
ASTHMA3     0
DIABETE4    0
SEX1        0
SMOKDAY2    0
STOPSMK2    0
SMOKE100    0
EXERANY2    0
SLEPTIM1    0
ALCDAY4     0
AVEDRNK3    0
MAXDRNKS    0
FRUIT2      0
VEGGIE1     0
EMPLOY1     0
_BMI5       0
_BMI5CAT    0
_SMOKER3    0
_AGE80      0
HTM4        0
WTKG3       0
INCOME3     0
EDUCA       0
_TOTINDA    0
FLUSHOT7    0
dtype: int64

In [44]:
# =============================================================================
# CELL 3b: Handle outliers using IQR method
# =============================================================================

def cap_outliers_iqr(dataframe, multiplier=1.5):
    """
    For each numeric column, cap values outside [Q1 - 1.5*IQR, Q3 + 1.5*IQR]
    to the boundary values (winsorization).
    """
    df_capped = dataframe.copy()
    outlier_report = []

    for col in df_capped.select_dtypes(include=[np.number]).columns:
        Q1 = df_capped[col].quantile(0.25)
        Q3 = df_capped[col].quantile(0.75)
        IQR = Q3 - Q1

        lower = Q1 - multiplier * IQR
        upper = Q3 + multiplier * IQR

        n_low = (df_capped[col] < lower).sum()
        n_high = (df_capped[col] > upper).sum()
        total = n_low + n_high

        if total > 0:
            df_capped[col] = df_capped[col].clip(lower=lower, upper=upper)
            outlier_report.append({
                'Column': col,
                'Lower': round(lower, 2),
                'Upper': round(upper, 2),
                'Outliers_Capped': total,
                'Pct': round(total / len(df_capped) * 100, 2)
            })

    report = pd.DataFrame(outlier_report)
    return df_capped, report

df, outlier_report = cap_outliers_iqr(df)

print("Outlier Report (IQR method):")
print(outlier_report.to_string(index=False))
print(f"\nShape after outlier handling: {df.shape}")

Outlier Report (IQR method):
Empty DataFrame
Columns: []
Index: []

Shape after outlier handling: (90522, 45)


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90522 entries, 0 to 90521
Data columns (total 45 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   _STATE    90522 non-null  float64
 1   IDATE     90522 non-null  int64  
 2   SEQNO     90522 non-null  int64  
 3   DISPCODE  90522 non-null  float64
 4   CTELENUM  90522 non-null  int64  
 5   NUMADULT  90522 non-null  int64  
 6   GENHLTH   90522 non-null  int64  
 7   PHYSHLTH  90522 non-null  int64  
 8   MENTHLTH  90522 non-null  float64
 9   POORHLTH  90522 non-null  int64  
 10  HLTHPLN1  90522 non-null  object 
 11  PERSDOC3  90522 non-null  object 
 12  MEDCOST1  90522 non-null  object 
 13  CHECKUP1  90522 non-null  object 
 14  BPHIGH6   90522 non-null  object 
 15  BPMEDS    90522 non-null  object 
 16  CHOLCHK3  90522 non-null  object 
 17  TOLDHI3   90522 non-null  object 
 18  CVDINFR4  90522 non-null  object 
 19  CVDCRHD4  90522 non-null  object 
 20  CVDSTRK3  90522 non-null  ob

In [13]:
df.duplicated().sum()

np.int64(0)

In [14]:
df.columns

Index(['_STATE', 'IDATE', 'SEQNO', 'DISPCODE', 'CTELENUM', 'NUMADULT',
       'GENHLTH', 'PHYSHLTH', 'MENTHLTH', 'POORHLTH', 'HLTHPLN1', 'PERSDOC3',
       'MEDCOST1', 'CHECKUP1', 'BPHIGH6', 'BPMEDS', 'CHOLCHK3', 'TOLDHI3',
       'CVDINFR4', 'CVDCRHD4', 'CVDSTRK3', 'ASTHMA3', 'DIABETE4', 'SEX1',
       'SMOKDAY2', 'STOPSMK2', 'SMOKE100', 'EXERANY2', 'SLEPTIM1', 'ALCDAY4',
       'AVEDRNK3', 'MAXDRNKS', 'FRUIT2', 'VEGGIE1', 'EMPLOY1', '_BMI5',
       '_BMI5CAT', '_SMOKER3', '_AGE80', 'HTM4', 'WTKG3', 'INCOME3', 'EDUCA',
       '_TOTINDA', 'FLUSHOT7'],
      dtype='object')

In [16]:
# =============================================================================
# Keep only Kenya-relevant columns (features + targets)
# =============================================================================

df_cleaned = df[['CVDINFR4','CVDCRHD4','SEX1','_AGE80','SMOKE100','_SMOKER3','EXERANY2','SLEPTIM1','BPHIGH6','TOLDHI3','DIABETE4','_BMI5','GENHLTH','PHYSHLTH','MENTHLTH','INCOME3','EDUCA']]



In [17]:
df_cleaned.columns

Index(['CVDINFR4', 'CVDCRHD4', 'SEX1', '_AGE80', 'SMOKE100', '_SMOKER3',
       'EXERANY2', 'SLEPTIM1', 'BPHIGH6', 'TOLDHI3', 'DIABETE4', '_BMI5',
       'GENHLTH', 'PHYSHLTH', 'MENTHLTH', 'INCOME3', 'EDUCA'],
      dtype='object')

# Exploratory Data Analysis